application/controllers/Notes.php

```diff
@@ -1,119 +1,317 @@
+                $suggested_title = strtoupper($this->callbook->get_plaincall($this->input->post('title', TRUE)));
+            }
+            $data['suggested_title'] = $suggested_title;
+            $data['page_title'] = __("Edit Note");
+            $this->load->view('interface_assets/header', $data);
+            $this->load->view('notes/edit');
+            $this->load->view('interface_assets/footer');
+        } else {
+            $category = $this->input->post('category', TRUE);
+            $title = $this->input->post('title', TRUE);
+            $content = $this->input->post('content', TRUE);
+            $local_time = $this->input->post('local_time', TRUE);
+            $note_id = $this->input->post('id', TRUE);
+            if ($category === 'Contacts') {
+                $title = strtoupper($this->callbook->get_plaincall($title));
+            }
+            $this->note->edit($note_id, $category, $title, $content, $local_time);
+            redirect('notes');
+        }
+    }
+
+    // API search for notes
+    function search() {
+        $this->load->model('note');
+        // Map and sanitize search parameters
+        $searchCriteria = $this->mapParameters();
+        // Get pagination and sorting parameters
+        $page = (int)$this->input->post('page', TRUE);
+        $per_page = (int)$this->input->post('per_page', TRUE);
+        $sort_col = $this->input->post('sort_col', TRUE);
+        $sort_dir = $this->input->post('sort_dir', TRUE);
+        if ($per_page < 1) $per_page = 15;
+        if ($page < 1) $page = 1;
+        // Get paginated, sorted notes
+        $result = $this->note->search_paginated($searchCriteria, $page, $per_page, $sort_col, $sort_dir);
+        $response = [
+            'notes' => $result['notes'],
+            'total' => $result['total']
+        ];
+        $this->output
+            ->set_content_type('application/json')
+            ->set_output(json_encode($response));
+    }
+
+    // Map and sanitize search parameters
+    function mapParameters() {
+        return array(
+            'cat' => $this->input->post('cat', TRUE),
+            'search' => $this->input->post('search', TRUE)
+        );
+    }
 
+    // Delete a note
+    function delete($id = null) {
 		$this->load->model('note');
-		$this->note->delete($clean_id);
-		
-		redirect('notes');
-	}
-}
\ No newline at end of file
+
+        $clean_id = $this->security->xss_clean($id);
+        // Validate note ID and ownership
+        if (!is_numeric($clean_id) || !$this->note->belongs_to_user($clean_id, $this->session->userdata('user_id'))) {
+            show_404();
+        }
+        $this->note->delete($clean_id);
+        redirect('notes');
+    }
+
+    // Duplicate a note by ID for the logged-in user, appending #[timestamp] to title
+    public function duplicate($id = null) {
+        $this->load->model('note');
+        $clean_id = $this->security->xss_clean($id);
+        if (!is_numeric($clean_id)) {
```

Comment:
We could make these check persistent just as the others aswell


```suggestion
        if (!is_numeric($clean_id) || !$this->note->belongs_to_user($clean_id, $this->session->userdata('user_id'))) {
```